solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1731,6 +1738,19 @@ private void getDocListC(QueryResult qr, QueryCommand cmd) throws IOException {
     }
   }
 
+  private Relation populateScoresIfNeeded(
+      QueryCommand cmd, boolean needScores, TopDocs topDocs, Query query, ScoreMode scoreModeUsed)
+      throws IOException {
+    if (cmd.getSort() != null && !(cmd.getQuery() instanceof RankQuery) && needScores) {
+      TopFieldCollector.populateScores(topDocs.scoreDocs, this, query);
+    }
+    if (scoreModeUsed == ScoreMode.COMPLETE || scoreModeUsed == ScoreMode.COMPLETE_NO_SCORES) {
+      return TotalHits.Relation.EQUAL_TO;
+    } else {
+      return topDocs.totalHits.relation;
+    }
+  }
+
```

Comment:
Proposing to factor out a method like this -- see #2280 -- which can then be called from all four places. I'm happy to resolve any merge conflicts that may arise here if #2280 is merged first.